datalad/support/annexrepo.py

```diff
@@ -776,8 +778,39 @@ def get_remotes(self,
         remotes : list of str
           List of names of the remotes
         """
+
         remotes = super(AnnexRepo, self).get_remotes(with_urls_only=with_urls_only)
 
+        # Note: the above seems fine for git remotes and special remotes, that show up in configs.
+        # However, this is insufficient to discover all annex special remotes.
```

Comment:
> Doesn't the presence of an external special remote is in .git/config come down to whether it's enabled?

I do think this is correct. I'm just not entirely sure whether there's more to it, so I didn't dare to phrase it that way ;-)

> So, as I understand it, this change would come with two behavioral changes:
>
>  - get_remotes() would now report on special remotes that are not enabled.
>
>  - it would also now include regular remotes that git-annex is aware of, but that aren't hooked up in the current repository as a git remote.
>
> Both of those seem problematic if any of the current callers assume they're getting operational remotes.

This is my intent indeed. I want to know about the remotes annex knows. I can see how this can be problematic, though. However, the assumption of "operational remotes" pretty much arises from the notion of git remotes, I guess. It's only annex special remotes that would introduce the distinction between enabled/not-enabled remotes. So, if this method is supposed to report on special remotes at all (which it clearly is, since we have switch to explicitly exclude special remotes), it should not pretend that not (yet) enabled remotes aren't there at all, but be able to capture the entire notion of annex remotes.

I'd be fine with introducing a switch to decide whether or not to query for those "non-operational (special) remotes" and default to False. Objection to that solution, @kyleam ?

Reply:
> > So, as I understand it, this change would come with two behavioral changes:
> > 
> > * get_remotes() would now report on special remotes that are not enabled.
> > * it would also now include regular remotes that git-annex is aware of, but that aren't hooked up in the current repository as a git remote.
> > 
> > Both of those seem problematic if any of the current callers assume they're getting operational remotes.
> 
> This is my intent indeed. I want to know about the remotes annex knows. I can see how this can be problematic, though. However, the assumption of "operational remotes" pretty much arises from the notion of git remotes, I guess. It's only annex special remotes that would introduce the distinction between enabled/not-enabled remotes.

Hmm I probably shouldn't have muddied the waters saying "operational".  By that, I was trying to cover both the special remotes and regular git remotes that I mentioned in the second bullet above.

  * A non-operational special remote would be one that isn't enabled (either through `git annex enableremote` or, in the case of `autoenable`, `git annex init`).

  * A non-operational normal git remote would be one that `git annex info` lists but that is not configured in that clone as a git remote.  In the case of a `datalad install`, this would commonly be a machine that the distributor has access to but the cloner does not.

> So, if this method is supposed to report on special remotes at all (which it clearly is, since we have switch to explicitly exclude special remotes), it should not pretend that not (yet) enabled remotes aren't there at all, but be able to capture the entire notion of annex remotes.

I suppose it depends on what the caller wants to do with the remotes.  If they intend to use them (rather than, say, tell the user about all known remotes), then I could see wanting only special remotes that are enabled and git remotes that are configured in the current repo.  To use the `datalad install -d. ///` example, it's not useful for a datalad command to try to work with a remote "yoh@smaug:/mnt/btrfs/datasets/datalad/crawl" (unless that user happens to be Yarik :>).

> I'd be fine with introducing a switch to decide whether or not to query for those "non-operational (special) remotes" and default to False.

For the reasons above, I don't agree with your "(special)" :)

> Objection to that solution, @kyleam ?

Adding a compatibility flag that keeps the old behavior addresses my (maybe unfounded) concern about breaking the current callers.  Thanks.

And a thought rather than an objection: If those inaccessible normal git remotes aren't of interest, I wonder whether `get_special_remotes` (which gets its info from git-annex:remote.log) would work well enough for your purposes.

https://github.com/datalad/datalad/blob/71a978987a464b4d00b7939434e29117d6a20f2f/datalad/support/annexrepo.py#L811-L813

Reply:
> For the reasons above, I don't agree with your "(special)" :)

I think we do agree. ;-)
The parenthesis were meant to make it "optional". So, I do want to include not only special remotes but also git remotes. Hell, we need a name for those. I tried "annex remotes" but that didn't work obviously. :-D

> And a thought rather than an objection: If those inaccessible normal git remotes aren't of interest, I wonder whether get_special_remotes (which gets its info from git-annex:remote.log) would work well enough for your purposes.

True, but in the case I stumbled upon this, I do need both.